shiny/_template_utils.py

```diff
@@ -0,0 +1,166 @@
+import shutil
+import sys
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+import questionary
+from questionary import Choice
+
+from ._custom_component_template_questions import (
+    install_js_dependencies,
+    isValidName,
+    updateComponentNameInTemplate,
+)
+
+
+def template_query(question_state: Optional[str] = "start"):
+    """
+    This will initiate a CLI query which will ask the user which template they would like.
+    If called without arguments this function will start from the top level and ask which
+    type of template the user would like.
+
+    You can also specify a question state to return to another level. For example if you
+    were at level 5 of a question chain and wanted to return to level 4.
+    This is not that useful currently because we only have two levels of questions.
+
+
+    :param question_state: The question state you would like to return to. Currently, the options are:
+        "cancel": Cancel the operation and exit.
+        "js-component": Start the questions for creating a custom JavaScript component.
+    """
+    if question_state is None:
+        template = questionary.select(
+            "Which template would you like to use?:",
+            choices=choicesArray(
+                [
+                    ("Basic App", "basic-app"),
+                    ("Express app", "express"),
+                    ("Dashboard", "dashboard"),
+                    ("Multi-page app with modules", "multi-page"),
+                    ("Custom JavaScript Component", "js-component"),
+                    ("Cancel", "cancel"),
+                ]
+            ),
+        ).ask()
+    else:
+        template = question_state
+
+    # Define the control flow for the top level menu
+    # This is simple now but will get more complicated as we add more templates
+    flow_dispatch = {
+        "cancel": lambda: sys.exit(),
+        "js-component": lambda: jsComponentQuestions(),
+        "default": lambda: appTemplateQuestions(template),
+    }
+
+    # Call the appropriate function based on the user's choice
+    flow_dispatch.get(template, flow_dispatch["default"])()
```

Comment:
If we don't go the way I suggested, at the very least, I don't like `"default"` being one of the `flow_dispatch` keys as it will never be the value of `template`.

```suggestion
    # Define the control flow for the top level menu
    # This is simple now but will get more complicated as we add more templates
    flow_dispatch = {
        "cancel": lambda: sys.exit(),
        "js-component": lambda: jsComponentQuestions(),
    }

    # Call the appropriate function based on the user's choice
    flow_dispatch.get(template, lambda: appTemplateQuestions(template))()
```

Reply:
Definitely. Was a poor thought out key and also indicative of my desire to strong-arm a switch statement into a language that doesnt have one.